src/tap/grpc/server.rs

```diff
@@ -255,24 +287,71 @@ impl iface::Tap for Tap {
                     base: shared.base_id,
                     stream: next_id as u64,
                 };
-                Some((id, shared.events_tx.clone()))
-            } else {
-                None
+                // Is HTTP data being extracted from the request, and should
+                // headers be included?
+                match shared.extract {
+                    ExtractKind::Http { headers } => {
+                        return Some((id, shared.events_tx.clone(), headers))
+                    }
+                }
             }
+            None
         })?;
```

Comment:
tioli: This block may be clearer with more eager failure (there's basically no need for the `and_then`, right? we can just do `let shared = self.shared.upgrade()?;`